In [1]:
import numpy as np
import pandas as pd

import statsmodels.api as sm
import statsmodels.formula.api as smf

# Bài 8.5.
Khảo sát tại 12 công ty trong một khu vực bán hàng về <em>doanh thu (Y)</em>, <em>chi phí quảng cáo (X<sub>1</sub>)</em> và <em>lương nhân viên tiếp thị (X<sub>2</sub>)</em> có số liệu sau:  
Y = [126, 148, 105, 162, 101, 175, 160, 127, 138, 143, 158, 137]  
X<sub>1</sub> = [17, 23, 18, 22, 14, 24, 23, 15, 16, 21, 22, 13]  
X<sub>2</sub> = [11, 14, 9, 16, 9, 17, 15, 11, 12, 14, 15, 13] 

<ol>
    <li>Hãy giả sử mối quan hệ giữa Y với X<sub>1</sub> và X<sub>2</sub> có thể biểu diễn bằng hàm hồi quy tuyến tính. Hãy ước lượng hàm này?</li>
    <li>Giải thích ý nghĩa của các hệ số hồi quy về mặt lý thuyết và thống kê?</li>
    <li>Tính hệ số xác định bội và hệ số xác định bội điều chỉnh và cho biết ý nghĩa?</li>
    <li>Phải chăng cả hai biến X<sub>1</sub> và X<sub>2</sub> đều không ảnh hưởng đến Y?</li>
    <li>Để dự báo doanh thu ta nên dùng hàm nào trong các hàm sau đây: 
        <ul>
            <li>y<sub>i</sub> = B<sub>0</sub> + B<sub>1</sub>x<sub>1i</sub> + e<sub>i</sub></li>
            <li>y<sub>i</sub> = B<sub>0</sub> + B<sub>2</sub>x<sub>2i</sub> + e<sub>i</sub></li>
            <li>y<sub>i</sub> = B<sub>0</sub> + B<sub>1</sub>x<sub>1i</sub> + B<sub>2</sub>x<sub>2i</sub> + e<sub>i</sub></li>
        </ul>
    </li>
    <li>Dự báo doanh thu trung bình của một công ty có chi phí quảng cáo là 23 triệu đông/tháng và lương của nhân viên tiếp thị là 15 triệu đồng/tháng với mức ý nghĩa 5%?</li>
</ol>

In [4]:
Y = [126, 148, 105, 162, 101, 175, 160, 127, 138, 143, 158, 137]
X1 = [17, 23, 18, 22, 14, 24, 23, 15, 16, 21, 22, 13]
X2 = [11, 14, 9, 16, 9, 17, 15, 11, 12, 14, 15, 13]

df = pd.DataFrame.from_dict({"X1": X1, "X2": X2, "Y": Y})
df

X1  X2    Y
0   17  11  126
1   23  14  148
2   18   9  105
3   22  16  162
4   14   9  101
5   24  17  175
6   23  15  160
7   15  11  127
8   16  12  138
9   21  14  143
10  22  15  158
11  13  13  137

## 1. Hãy giả sử mối quan hệ giữa Y với X<sub>1</sub> và X<sub>2</sub> có thể biểu diễn bằng hàm hồi quy tuyến tính. Hãy ước lượng hàm này?

In [5]:
results = smf.ols("Y ~ X1 + X2", data=df).fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.972
Model:                            OLS   Adj. R-squared:                  0.966
Method:                 Least Squares   F-statistic:                     157.9
Date:                Sat, 09 Jul 2022   Prob (F-statistic):           9.81e-08
Time:                        19:35:45   Log-Likelihood:                -32.416
No. Observations:                  12   AIC:                             70.83
Df Residuals:                       9   BIC:                             72.29
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     29.6619      6.629      4.474      0.0

/Users/phu/opt/anaconda3/envs/EdaK31/lib/python3.7/site-packages/scipy/stats/stats.py:1542: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  "anyway, n=%i" % int(n))


Trả lời: Có thể ước lượng mối quan hệ bằng hàm:  
y<sub>i</sub> = B<sub>0</sub> + B<sub>1</sub>x<sub>1i</sub> + B<sub>2</sub>x<sub>2i</sub>  
y<sub>i</sub> = 29.66 + 0.0023x<sub>1i</sub> + 8.4842x<sub>2i</sub>

## 2. Giải thích ý nghĩa của các hệ số hồi quy về mặt lý thuyết và thống kê?

Trả lời: Các hệ số hồi quy cho thấy doanh thu của một công ty bán hàng ít phụ thuộc vào chi phí quảng cáo mà phụ thuộc nhiều vào lương của nhân viên tiếp thị.  Nếu lương của nhân viên tiếp thị tăng lên 1 triệu đồng thì doanh thu tăng trung bình 8.4842 triệu đồng.

## 3. Tính hệ số xác định bội và hệ số xác định bội điều chỉnh và cho biết ý nghĩa?
Hệ số xác định bội R<sup>2</sup>: 0.972  
Hệ số xác định bội hiệu chỉnh R<sup>2</sup><sub>adj</sub>: 0.966  

Hệ số xác định R<sup>2</sup> cho biết mô hình hồi quy tuyến tính này phù hợp với tập dữ liệu ở mức 97.2%. Nói cách khác, 97.2% độ biến thiên của biến phụ thuộc được giải thích bởi các biến độc lập.  
Hệ số xác định R<sup>2</sup><sub>adj</sub> để hiệu chỉnh R<sup>2</sup> theo bậc tự do. Vì nếu chỉ sử dụng R<sup>2</sup>, khi đưa thêm nhiều biến vào mô hình, mặc dù chưa xác định biến đưa vào có ý nghĩa hay không thì giá trị R<sup>2</sup> sẽ tăng. Điều này dẫn đến R<sup>2</sup> bản chất không làm rõ được tầm quan trọng của biến đưa vào.

## 4. Phải chăng cả hai biến X<sub>1</sub> và X<sub>2</sub> đều không ảnh hưởng đến Y?
Trả lời: X<sub>2</sub> có ảnh hưởng lớn đến Y còn  X<sub>1</sub> không có ý nghĩa về mặt thống kê.

## 5. Để dự báo doanh thu ta nên dùng hàm nào trong các hàm sau đây: 
<ul>
    <li>y<sub>i</sub> = B<sub>0</sub> + B<sub>1</sub>x<sub>1i</sub> + e<sub>i</sub></li>
    <li>y<sub>i</sub> = B<sub>0</sub> + B<sub>2</sub>x<sub>2i</sub> + e<sub>i</sub></li>
    <li>y<sub>i</sub> = B<sub>0</sub> + B<sub>1</sub>x<sub>1i</sub> + B<sub>2</sub>x<sub>2i</sub> + e<sub>i</sub></li>
</ul>

In [6]:
results = smf.ols("Y ~ X1", data=df).fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.583
Model:                            OLS   Adj. R-squared:                  0.541
Method:                 Least Squares   F-statistic:                     13.98
Date:                Sat, 09 Jul 2022   Prob (F-statistic):            0.00385
Time:                        22:26:21   Log-Likelihood:                -48.685
No. Observations:                  12   AIC:                             101.4
Df Residuals:                      10   BIC:                             102.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     56.5118     22.761      2.483      0.0

/Users/phu/opt/anaconda3/envs/EdaK31/lib/python3.7/site-packages/scipy/stats/stats.py:1542: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  "anyway, n=%i" % int(n))


In [7]:
results = smf.ols("Y ~ X2", data=df).fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.972
Model:                            OLS   Adj. R-squared:                  0.970
Method:                 Least Squares   F-statistic:                     350.9
Date:                Sat, 09 Jul 2022   Prob (F-statistic):           4.07e-09
Time:                        22:26:28   Log-Likelihood:                -32.416
No. Observations:                  12   AIC:                             68.83
Df Residuals:                      10   BIC:                             69.80
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     29.6711      5.999      4.946      0.0

/Users/phu/opt/anaconda3/envs/EdaK31/lib/python3.7/site-packages/scipy/stats/stats.py:1542: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  "anyway, n=%i" % int(n))


In [8]:
results = smf.ols("Y ~ X1 + X2", data=df).fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.972
Model:                            OLS   Adj. R-squared:                  0.966
Method:                 Least Squares   F-statistic:                     157.9
Date:                Sat, 09 Jul 2022   Prob (F-statistic):           9.81e-08
Time:                        22:26:54   Log-Likelihood:                -32.416
No. Observations:                  12   AIC:                             70.83
Df Residuals:                       9   BIC:                             72.29
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     29.6619      6.629      4.474      0.0

/Users/phu/opt/anaconda3/envs/EdaK31/lib/python3.7/site-packages/scipy/stats/stats.py:1542: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  "anyway, n=%i" % int(n))


Trả lời: Ta thấy hàm  
- y<sub>i</sub> = B<sub>0</sub> + B<sub>2</sub>x<sub>2i</sub> + e<sub>i</sub>  
y<sub>i</sub> = 29.6711 + 8.4868*x<sub>2i</sub> + e<sub>i</sub>  

cho ra R<sup>2</sup><sub>adj</sub> tốt nhất (0.970) nên ta nên dùng hàm này để dự báo doanh thu.

## 6. Dự báo doanh thu trung bình của một công ty có chi phí quảng cáo là 23 triệu đông/tháng và lương của nhân viên tiếp thị là 15 triệu đồng/tháng với mức ý nghĩa 5%?

Trả lời: Với mức ý nghĩa 5%, dựa vào kết quả ở trên, ta có thể loại bỏ được biến chi phí quảng cáo. Như vậy doanh thu dự báo sẽ là:  
y = 29.6711 + 8.4868 * 15 = 156 (triệu đồng).